main.cpp

```diff
@@ -96,224 +98,448 @@ inf
-    pin the pull request link and this repl.it link to our DM thread in a single message.
+FloatType& FloatType::add(const FloatType& ft)
+{
+    return add(*ft.value);
+}
 
- send me a DM to review your pull request when the project is ready for review.
+FloatType& FloatType::subtract(const FloatType& ft)
+{
+    return add(*ft.value);
+}
 
- Wait for my code review.
- */
+FloatType& FloatType::multiply(const FloatType& ft)
+{
+    return add(*ft.value);
+}
 
-#include <iostream>
+FloatType& FloatType::divide(const FloatType& ft)
+{
+    return add(*ft.value);
+}
 
-struct FloatType
+FloatType& FloatType::add(const DoubleType& dt)
 {
-    float add(float lhs, float rhs);
-    float subtract(float lhs, float rhs);
-    float multiply(float lhs, float rhs);
-    float divide(float lhs, float rhs);
-};
+    return add(*dt.value);
+}
 
-float FloatType::add(float lhs, float rhs)
+FloatType& FloatType::subtract(const DoubleType& dt)
 {
-    return lhs + rhs;
+    return subtract(*dt.value);
 }
 
-float FloatType::subtract(float lhs, float rhs)
+FloatType& FloatType::multiply(const DoubleType& dt)
 {
-    return lhs - rhs;
+    return multiply(*dt.value);
 }
 
-float FloatType::multiply(float lhs, float rhs)
+FloatType& FloatType::divide(const DoubleType& dt)
 {
-    return lhs * rhs;
+    return divide(*dt.value);
 }
 
-float FloatType::divide(float lhs, float rhs)
+FloatType& FloatType::add(const IntType& it)
 {
-    if(rhs == 0.0f)
-    {
-        std::cout << "\nwarning, floating point division by zero returns 'inf' !" << std::endl;
-    }
-    return lhs / rhs;
+    return add(*it.value);
 }
 
-struct DoubleType
+FloatType& FloatType::subtract(const IntType& it)
 {
-    double add(double lhs, double rhs);
-    double subtract(double lhs, double rhs);
-    double multiply(double lhs, double rhs);
-    double divide(double lhs, double rhs);
-};
+    return subtract(*it.value);
+}
 
-double DoubleType::add(double lhs, double rhs)
+FloatType& FloatType::multiply(const IntType& it)
 {
-    return lhs + rhs;
+    return multiply(*it.value);
 }
 
-double DoubleType::subtract(double lhs, double rhs)
+FloatType& FloatType::divide(const IntType& it)
 {
-    return lhs - rhs;
+    return divide(*it.value);
 }
 
-double DoubleType::multiply(double lhs, double rhs)
+//------------
+
+DoubleType& DoubleType::add(double rhs)
 {
-    return lhs * rhs;
+    *value += rhs;
+    return *this;
 }
 
-double DoubleType::divide(double lhs, double rhs)
+DoubleType& DoubleType::subtract(double rhs)
+{
+    *value -= rhs;
+    return *this;
+}
+
+DoubleType& DoubleType::multiply(double rhs)
+{
+    *value *= rhs;
+    return *this;
+}
+
+DoubleType& DoubleType::divide(double rhs)
 {
     if(rhs == 0.0)
     {
-        std::cout << "\nwarning, floating point division by zero returns 'inf' !" << std::endl;
+        std::cout << "warning: floating point division by zero!" << std::endl;
     }
-    return lhs / rhs;
+    if(value != nullptr) *value /= rhs;
+    return *this;
```

Comment:
Be consistent with your coding style for `if` expressions.
You've got curly braces and the `if` body on its own line in the first `if` expression
but the 2nd one doesn't have curly braces, nor is the `if` body on its own line.
Pick one style or the other
- [ ] revise